### ISSUE TITLE
Tor identity isolation is broken

### ISSUE THREAD

**Issue by iguanaaaaaa** [sparrowwallet/sparrow#1328]
Good afternoon! I was excited to find an accessible and affordable CoinJoin implementation in Sparrow, so I looked into how it is implemented. I focused on the network layer, since math isn't my strongest suit. I found a few issues which could lead to the compromise of the CoinJoin outputs' anonymity by a malicious coordinator server.

## Changing Tor identities
As far as I can tell, the only method of isolating Tor identities employed by Sparrow is sending NEWNYM to the Tor control port:
https://github.com/sparrowwallet/sparrow/blob/f3c44e6f3e49af6e79e2f2862a70552ea5c2616c/src/main/java/com/sparrowwallet/sparrow/net/TorUtils.java#L15-L30

There're two implementations of (globally) rotating identities that use stream isolation (through supplying different SOCKS credentials), however those seem to be dead code:
https://github.com/sparrowwallet/nightjar/blob/08fd60cd5256ce4842701c6c179c18f997fd252f/src/main/java/com/sparrowwallet/nightjar/tor/WhirlpoolTorClientService.java#L13-L19
https://github.com/sparrowwallet/sparrow/blob/f3c44e6f3e49af6e79e2f2862a70552ea5c2616c/src/main/java/com/sparrowwallet/sparrow/AppServices.java#L489-L496

The former is never called, it's overriden by the NEWNYM version. The latter is always called with the default circuit ID, rendering it useless.

### Control port authentication
When authenticating to an external Tor daemon, which is the only option available to local Bitcoin Core users, Sparrow uses an empty password:
https://github.com/sparrowwallet/sparrow/blob/f3c44e6f3e49af6e79e2f2862a70552ea5c2616c/src/main/java/com/sparrowwallet/sparrow/net/TorUtils.java#L32-L36

Authentication failures are logged, but are never fatal, and no notification is presented to the user. Which effectively means that switching identities will silently fail for anyone using Tor with control port authentication enabled. It is the default configuration used in Tor Browser, which is in turn recommended in [Sparrow docs](https://www.sparrowwallet.com/docs/mixing-whirlpool.html#starting-the-mix).

I would suggest implementing cookie/safecookie authentication, which would instantly make this functionality accessible to everyone using Tor Browser, no setup required. I would also suggest adding an option to require successful identity change before proceeding with further Whirlpool rounds and/or transaction broadcasts. When this option is switched on and control port authentication fails, all external network connections would be shut down immediately. A dialog window suggesting further actions would be presented to the user.

### Coordinator server communication
When unathenticated Tor control port access is possible, it is used to change identities between successful Whirlpool mixes:
https://github.com/sparrowwallet/nightjar/blob/08fd60cd5256ce4842701c6c179c18f997fd252f/src/main/java/com/samourai/whirlpool/client/wallet/WhirlpoolWallet.java#L623

Before broadcasting a transaction:
https://github.com/sparrowwallet/sparrow/blob/f3c44e6f3e49af6e79e2f2862a70552ea5c2616c/src/main/java/com/sparrowwallet/sparrow/net/BroadcastSource.java#L145

And on connection failure, which I'm assuming has more to do with network stability than anonymity:
https://github.com/sparrowwallet/nightjar/blob/08fd60cd5256ce4842701c6c179c18f997fd252f/src/main/java/com/samourai/whirlpool/client/mix/dialog/MixSession.java#L266
https://github.com/sparrowwallet/sparrow/blob/f3c44e6f3e49af6e79e2f2862a70552ea5c2616c/src/main/java/com/sparrowwallet/sparrow/whirlpool/WhirlpoolServices.java#L126

Crucially, it doesn't happen when it is needed most: during Whirlpool output registration. The most important detail of Chaumian CoinJoin is that the outputs need to be registered using a separate identity, which could not in any way be linked to the identity used for registering the inputs. Otherwise, the use of Chaum's blind signatures is pointless, as you have to trust the server not to store the logs linking those identities together. At that point, it becomes equivalent to using a trusted coordinator server with a no-logs policy. It can't steal the coins, but if it violates its policy, then the anonymity is compromised.

All communication with the coordinator server seems to happen through the same Tor circuit, using separate Jetty HTTP client instances based on their "HTTP usage":
https://github.com/sparrowwallet/nightjar/blob/08fd60cd5256ce4842701c6c179c18f997fd252f/src/main/java/com/sparrowwallet/nightjar/http/JavaHttpClientService.java#L80-L111

I'm not sure if the use of separate Jetty instances is intended to boost anonymity or to simply avoid blocking the thread with an active WebSocket. However I believe that the Tor daemon is able to identify which circuit the connection comes from, connecting the registered inputs with the outputs. Either way, the current `JavaHttpClientService` is a good starting point to at least implement stream isolation, which shouldn't be a big change.

As far as I understand, this is the same implementation as used by Samourai. Whirlpool seems tricky to implement correctly, and the usefulness of doing it right is questionable if the client with the most liquidity doesn't. However this could change as more and more wallets start to support Whirlpool.

### Persistent HTTP connections
Even after stream isolation is implemented and Tor control port authentication is fixed, special care needs to be taken to avoid creating persistent HTTP connections. Jetty [uses them by default](https://github.com/jetty/jetty.project/issues/1280#issuecomment-273719444) for HTTP/1.1 and HTTP/2.0, and HTTP/1.0 if the `Connection: keep-alive` header is specified. Judging by Sparrow logs, the current coordinator server appears to set it (look for `502 BAD_GATEWAY Bad Gateway`).

A persistent HTTP connection will re-use the same TCP socket for consecutive requests instead of creating a new one. This is important, because signalling NEWNYM [does not force Tor to close existing TCP streams](https://tor.stackexchange.com/a/10618), it only affects the newly created ones. This could lead to consecutive mixes being linked together by the coordinator server.

Correct me if I'm wrong, but I could not find any code that would prevent Jetty from using persistent HTTP connections. The restart method of `JavaHttpClientService` is never called.

### TLS session tickets
Along with persistent HTTP connections, the [Tor Browser design draft](https://2019.www.torproject.org/projects/torbrowser/design/) mentions TLS session tickets as a potential deanonymization vector. I don't have deep enough knowledge of the Java internals to understand whether this applies to Sparrow, but Java 17 [appears](https://docs.oracle.com/en/java/javase/17/security/transport-layer-security-tls-protocol-overview.html#GUID-F32E1E87-F6D2-47E6-ABF4-EF219E6A1EB7) to use TLS session tickets for TLS 1.3 connections by default.

### Stream isolation vs NEWNYM
While researching stream isolation, I came across an interesting Stack Exchange question: [Signal NEWNYM vs Stream isolation](https://tor.stackexchange.com/questions/16398/signal-newnym-vs-stream-isolation). According to the answer, the hidden service might be able to identify separate Tor circuits coming from the same Tor instance. Which would mean that a correct implementation of Chaumian CoinJoin requires either the use of two unrelated instances of the Tor daemon, or one Tor daemon switching identities before and after output registration (but this means no WebSockets).

I later realized that the question was both asked and answered by @nopara73, however the information from the answer seems to never have been incorporated in Wasabi code. (**Update**: Wasabi consider the potential attack to be far-fetched, so they went with stream isolation instead: https://github.com/zkSNACKs/WalletWasabi/discussions/12679.)

Could you confirm whether or not my findings are accurate?

**Comment by craigraw** [comment]
Thanks for your detailed notes. 

> There're two implementations of (globally) rotating identities that use stream isolation

There are actually three - the implementation has changed somewhat over time: The switching of Tor identities in the Whirlpool client is handled by [SparrowTorClientService.java](https://github.com/sparrowwallet/sparrow/blob/9d0c35bc750091abdca2769b4ff6fde819d7503e/src/main/java/com/sparrowwallet/sparrow/whirlpool/tor/SparrowTorClientService.java). This is passed in as a config parameter to the whirlpool client, and used as necessary during mixing.

> Crucially, it doesn't happen when it is needed most: during Whirlpool output registration.

I haven't checked this recently, but I believe this is incorrect - the Whirlpool client follows the Zerolink protocol and refreshes the Tor circuit after registration.

I will comment on the rest of your findings in due course, but I should note the transport layer in Whirlpool is about to change fairly dramatically with decentralisation introduced using Soroban. You are welcome to review that code in the `whirlpool-1.0.0` branch.

**Comment by iguanaaaaaa** [comment]
> The switching of Tor identities in the Whirlpool client is handled by [SparrowTorClientService.java](https://github.com/sparrowwallet/sparrow/blob/9d0c35bc750091abdca2769b4ff6fde819d7503e/src/main/java/com/sparrowwallet/sparrow/whirlpool/tor/SparrowTorClientService.java). This is passed in as a config parameter to the whirlpool client, and used as necessary during mixing.

It uses the TorUtils.java method that I quoted, so I didn't think it was worth a separate mention.

> I haven't checked this recently, but I believe this is incorrect - the Whirlpool client follows the Zerolink protocol and refreshes the Tor circuit after registration.

If you search for `changeIdentity` in Nightjar (I like the bird species naming pattern, by the way!), it returns two results where this method is actually called, both linked in the issue.

> the transport layer in Whirlpool is about to change fairly dramatically with decentralisation introduced using Soroban

Interesting. At a glance, Soroban seems to be a centralized message-passing service (as of soroban-client-java 1.0.0-beta9 in the Samourai repo). If that's the case, it doesn't change much. I could be wrong though. I'll wait for an update to the Nightjar to take a better look.

**Comment by craigraw** [comment]
> I would suggest implementing cookie/safecookie authentication, which would instantly make this functionality accessible to everyone using Tor Browser, no setup required.

I have implemented support for cookie authentication in 08ec158d, and tested this is working with Tor Browser. This should cover most common situations.

>  I would also suggest adding an option to require successful identity change before proceeding with further Whirlpool rounds and/or transaction broadcasts.

I am specifically trying to avoid adding too much complexity to Whirlpool configuration, so I'd prefer to keep working to avoid such a setting.

> Crucially, it doesn't happen when it is needed most: during Whirlpool output registration.

I decided to focus on the new "Whirlpool-over-Soroban" implementation in the `whirlpool-1.0.0` branch when looking at this. Bear in mind I am not the author of this client, rather, the Samourai team have developed it. Here are their notes on identity switching: 

> "The Tor identity switch for REGISTER_OUTPUT was replaced with a Soroban identity switch: https://code.samourai.io/whirlpool/whirlpool-client/-/blob/dex-wp/src/main/java/com/samourai/whirlpool/client/mix/MixClient.java?ref_type=heads#L276. The output is pushed to a random Soroban node with a new Soroban identity, and the payload is encrypted between coordinator and mixing users. The coordinator doesn't see clients IP anymore (as it's pulling from Soroban), so there is very little chance to get the mix deanonymized as it would require a man-of-the-middle attack over Soroban nodes." 

That said, the Tor identity switch will be added back for additional privacy.

> special care needs to be taken to avoid creating persistent HTTP connections

I have confirmed that `Connection: keep-alive` is not specified. The client uses HTTP/1.1 POST calls with JSON-RPC payloads, for example:
```
Request:
POST /rpc HTTP/1.1
Accept-Encoding: gzip
Content-Type: application/json
Host: redacted
Content-Length: 194

{"method":"directory.List","id":1,"jsonrpc":"2.0","params":[{"Name":"TESTNET/WHIRLPOOL/1.0/REGISTER_INPUT/0.001btc/true/REPLY/redacted"}]}

Response:
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Vary: Origin
X-Content-Type-Options: nosniff
Date: Tue, 26 Mar 2024 13:49:20 GMT
Content-Length: 177

{"result":{"Name":"TESTNET/WHIRLPOOL/1.0/REGISTER_INPUT/0.001btc/true/REPLY/redacted","Entries":[]},"error":null,"id":1}
```

>  At a glance, Soroban seems to be a centralized message-passing service

Not quite. It's designed as a decentralized message passing service. The number of Soroban nodes is still small, but I expect it to grow strongly with integration on Ronin Dojo.

**Comment by iguanaaaaaa** [comment]
> I have implemented support for cookie authentication in [08ec158](https://github.com/sparrowwallet/sparrow/commit/08ec158d19beb4b973baa6284075be380f39990b), and tested this is working with Tor Browser. This should cover most common situations.

Cool, thank you!

> "The output is pushed to a random Soroban node with a new Soroban identity, and the payload is encrypted between coordinator and mixing users. The coordinator doesn't see clients IP anymore (as it's pulling from Soroban), so there is very little chance to get the mix deanonymized as it would require a man-of-the-middle attack over Soroban nodes."

Well, this raises more questions than it answers. How does it select the nodes? It appears that a few would be hardcoded, but those files are missing from the source repo. The link to `SorobanServerDex` [here](https://code.samourai.io/whirlpool/Whirlpool/-/blob/DexWp/ARCHITECTURE.md) is broken. No such file exists in either the Soroban or the Whirlpool repo (this type of thing is why I was going to wait for the Nightjar update). If Samourai control discovery, they don't have to MitM anyone. I suppose you aren't the person to address those questions to, though.

> I have confirmed that `Connection: keep-alive` is not specified. The client uses HTTP/1.1 POST calls with JSON-RPC payloads

The 502 endpoint must be different. The defaults still apply, i. e. HTTP/1.1 implies keep-alive. This likely won't matter with the decentralized Soroban however, depending on the implementation details.
